db/schema_tables.cc

```diff
@@ -1526,15 +1526,11 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
         // In order to avoid possible races we first create the tables and only then the views.
         // That way if a view seeks information about its base table it's guarantied to find it.
         co_await max_concurrent_for_each(tables_diff.created, max_concurrent, [&] (global_schema_ptr& gs) -> future<> {
-            co_await db.add_column_family_and_make_directory(gs);
+            co_await db.add_column_family_and_make_directory(gs, false);
```

Comment:
This patch doesn't seem to be a full solution to the problem:

> 
> This problem was demonstrated by materialised view dtests.
> A view is created on all nodes. On some nodes it will be created
> earlier than on others and the view rebuild process will start
> writing data to that view on other nodes, where mark_table_as_writable
> has not yet been called.

The table may have not yet been created on "other" node, and the update will fail on missing table. So it doesn't help that we mark it as ready for writes sooner, it may still not be soon enough. Another thing is that we shouldn't hit this problem because what should happen in this case is that such update should trigger schema pull on the replica due to an unsynced schema version and should proceed after schema in synced, which means that the table should be ready on all the shards by the time the update is attempted.

Maybe you hit https://github.com/scylladb/scylladb/issues/15048

Reply:
@tgrabiec re o_dsync, I don't see any reason to not use it if we know we want all writes to the file to be done in sync mode. 

The reason it is not sufficient to provide duarbility is that not all file systems support o_dsync, on these you need to additionally flush the write explicitly. But we don't run the commit log on such file systems. In all practical installations it is going to be enabled and give a significant performance boost.
It should always be set for schema commit log.

Reply:
> he table may have not yet been created on "other" node, and the update will fail on missing table. So it doesn't help that we mark it as ready for writes sooner, it may still not be soon enough. Another thing is that we shouldn't hit this problem because what should happen in this case is that such update should trigger schema pull on the replica due to an unsynced schema version and should proceed after schema in synced, which means that the table should be ready on all the shards by the time the update is attempted.

My reasoning around this approach:
* I managed to reproduce the views problem steadily on  the dtest `update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_garbage_members_from_group0_after_abort_decommission[leaving_token_ring-]`, this test failed consistently on my dev machine;
* The failures happen only with my commit where I trigger an internal_error when the `table.commitlog()` is accessed on the table, which has not yet been made writable through `table::mark_ready_for_writes`;
* In the failure backtrace I saw "Failure to apply mutations from <some ip>", where some ip != this node ip;
* No failures if we apply this commit;
* Looking at the `view_builder` code I saw a lot of `catch(no_such_column_family)` e.g. in `view_builder::on_create_view`, so I though it was expected when we build the view on another node and the current what doesn't have the view table yet, maybe rebuild will be retried later.

Reply:
Ok, I guess to avoid regressions we have to make this patch. Probably, the reason why it didn't fail before was that we mark the schema as synced when we create the view in add_column_family():
```
    schema->registry_entry()->mark_synced();
``` 
which is wrong (https://github.com/scylladb/scylladb/issues/14146) and we should move this mark out of there and do it after all shards have the table. Then this patch can be reverted.

Reply:
> @tgrabiec re o_dsync, I don't see any reason to not use it if we know we want all writes to the file to be done in sync mode.
> 
> The reason it is not sufficient to provide duarbility is that not all file systems support o_dsync, on these you need to additionally flush the write explicitly. But we don't run the commit log on such file systems. In all practical installations it is going to be enabled and give a significant performance boost. It should always be set for schema commit log.

Just curious, which file systems don't support O_DSYNC? I thought it was implemented purely in VFS.

Reply:
https://twitter.com/AlexMillerDB/status/1532280950779891712 recent relevant discussion about O_DSYNC on Darwin

Reply:
Reason not to use O_DSYNC in tests: https://github.com/scylladb/scylladb/pull/12542

Reply:
Perhaps a way out would be to enable O_DSYNC for schema commitlog but disable segment preallocation: https://github.com/scylladb/scylladb/issues/15349

Reply:
I think Scylla is on a safe side since it uses O_DIRECT. For buffered writes, things can go wrong in many other ways. This PostgreSQL page summarizes the issue: https://wiki.postgresql.org/wiki/Fsync_Errors

Reply:
In any case we don't do fsync-failure-testing as is suggested here, https://www.usenix.org/system/files/atc20-rebello.pdf

Reply:
What's wrong with disabling O_DSYNC for tests? they don't need crash recovery.

Reply:
And we don't care about Darwin (the OS).

Reply:
I expect that NFS emulates O_DSYNC by sending flushes after each write.

Reply:
> What's wrong with disabling O_DSYNC for tests? they don't need crash recovery.

Perhaps we'd like to test crash recovery. Especially with these changes around durability.